File: storage/insights.py
"""
Insight extraction from card chats.
Extracts learning insights incrementally using AI, merges with existing insights.
Uses backend /insights/extract endpoint for extraction calls.
"""
import hashlib
import json
import re

try:
    from ..utils.logging import get_logger
except ImportError:
    from utils.logging import get_logger
logger = get_logger(__name__)




def insight_hash(text):
    """Deterministic 8-char hash of insight text for seen-tracking."""
    return hashlib.md5(text.encode('utf-8')).hexdigest()[:8]


def compute_new_indices(insights, seen_hashes):
    """Return indices of insights whose text hash is not in seen_hashes."""
    return [
        i for i, ins in enumerate(insights)
        if insight_hash(ins.get('text', '')) not in seen_hashes
    ]


EXTRACTION_PROMPT = """Du extrahierst Lernerkenntnisse aus einem Chat über eine Anki-Lernkarte.

KARTE: {question}

BISHERIGE ERKENNTNISSE: {existing_insights}

CHAT:
{chat_messages}

FOKUS:
Erfasse was der NUTZER in diesem Chat gelernt, gefragt oder falsch verstanden hat.
Fasse NICHT den Tutor-Text zusammen — erfasse das Lernverhalten des Nutzers:

1. FRAGEN des Nutzers → zeigen Wissenslücken oder aktives Interesse (type: "learned")
2. FEHLER/VERWECHSLUNGEN → was der Nutzer falsch verstanden hat (type: "weakness")
3. UNSICHERHEITEN → wo der Nutzer nachfragen musste (type: "weakness")
4. SCHLÜSSELMOMENTE → was der Nutzer erst durch die Erklärung verstanden hat (type: "learned")

REGELN:
- Formuliere aus Nutzerperspektive: "Gefragt nach X", "Verwechslung: X vs Y", "Verstanden: X"
- Jede Frage des Nutzers ist eine potenzielle Erkenntnis (zeigt Interesse oder Wissenslücke)
- NUR kartenrelevante Erkenntnisse, kein Off-Topic/Smalltalk
- Merge mit bisherigen Erkenntnissen: Duplikate entfernen, max 10 Einträge
- Wenn keine neuen Erkenntnisse: bisherige unverändert zurückgeben
- NUR das JSON-Objekt ausgeben, KEIN anderer Text

BEISPIEL-OUTPUT:
{{"version":1,"insights":[{{"text":"Gefragt nach weiteren Ansätzen am Tuber ischiadicum","type":"learned"}},{{"text":"Verwechslung: M. pterygoideus lateralis vs. medialis","type":"weakness"}},{{"text":"Verstanden: Km ändert sich nur bei kompetitiver Hemmung","type":"learned"}}]}}"""


def _strip_html(text):
    """Remove HTML tags and collapse whitespace."""
    if not text:
        return ''
    text = re.sub(r'<[^>]+>', ' ', text)
    text = re.sub(r'\s+', ' ', text).strip()
    return text[:300]


def _filter_messages_for_extraction(messages):
    """Filter messages for extraction: include user + tutor + research, exclude Plusi + tool calls.

    Inclusion rules:
    - User messages: always included
    - Tutor/assistant responses: included (main learning content)
    - Research agent results: included (factual content relevant to learning)
    - Plusi messages: EXCLUDED (personality/emotional, not learning content)
    - Tool/function calls: EXCLUDED (internal mechanics)
    - Tool widget markers: EXCLUDED (UI signals)
    """
    filtered = []
    for m in messages:
        # Skip tool/function calls
        if m.get('is_function_call'):
            continue

        # Skip tool widget markers
        text = m.get('text', '')
        if text.startswith('[[TOOL') or text.startswith('[[LOADING'):
            continue

        sender = m.get('sender', m.get('from', ''))

        # Skip Plusi messages (subagent personality responses)
        if m.get('subagent') == 'plusi' or m.get('agent') == 'plusi':
            continue

        # Include user messages
        if sender == 'user':
            filtered.append(m)
            continue

        # Include assistant/bot/tutor messages
        if sender in ('assistant', 'bot'):
            filtered.append(m)
            continue

    return filtered


def _format_chat_for_extraction(messages):
    """Format filtered messages as compact text for the extraction prompt."""
    filtered = _filter_messages_for_extraction(messages)
    lines = []
    for m in filtered:
        sender = "User" if m.get('sender', m.get('from')) == 'user' else "Tutor"
        text = m.get('text', '')[:400]
        lines.append(f"{sender}: {text}")
    result = "\n".join(lines)
    return result[:3000]


def _count_user_messages(messages):
    """Count messages from the user."""
    return sum(1 for m in messages if m.get('sender', m.get('from')) == 'user')


def build_extraction_prompt(card_context, messages, existing_insights, performance_data=None):
    """Build the full extraction prompt."""
    question = card_context.get('frontField', '')
    if not question:
        question = _strip_html(card_context.get('question', ''))

    existing_str = json.dumps(existing_insights, ensure_ascii=False) if existing_insights.get('insights') else "Keine"
    chat_str = _format_chat_for_extraction(messages)

    return EXTRACTION_PROMPT.format(
        question=question[:300],
        existing_insights=existing_str,
        chat_messages=chat_str,
    )


def extract_insights_via_openrouter(prompt, api_key=''):
    """Extract insights via backend /insights/extract endpoint.

    Args:
        prompt: The extraction prompt (contains card context + chat messages).
        api_key: Unused (kept for backward compat).

    Returns the raw response text (JSON string) or raises on error.
    """
    try:
        from ..config import get_backend_url, get_auth_token
    except ImportError:
        from config import get_backend_url, get_auth_token

    backend_url = get_backend_url()
    auth_token = get_auth_token()

    if not backend_url or not auth_token:
        raise ValueError("Nicht authentifiziert")

    import requests
    headers = {
        'Authorization': 'Bearer %s' % auth_token,
        'Content-Type': 'application/json',
    }
    resp = requests.post('%s/insights/extract' % backend_url.rstrip('/'),
                         json={'messages': [{'role': 'user', 'content': prompt}]},
                         headers=headers, timeout=20)
    resp.raise_for_status()
    data = resp.json()

    logger.info("Insight extraction via backend completed")

    # Caller expects raw text for parse_extraction_response() to parse
    return json.dumps(data)


def parse_extraction_response(response_text):
    """Parse AI response into insights JSON. Returns None on failure."""
    try:
        text = response_text.strip()
        # Strip markdown code fences
        if '```json' in text:
            text = text.split('```json')[1].split('```')[0].strip()
        elif '```' in text:
            text = text.split('```')[1].split('```')[0].strip()

        # Try to find JSON object in text
        json_match = re.search(r'\{[\s\S]*"insights"\s*:\s*\[[\s\S]*\]\s*\}', text)
        if json_match:
            text = json_match.group(0)

        data = json.loads(text)

        if 'insights' not in data or not isinstance(data['insights'], list):
            return None

        data['version'] = data.get('version', 1)

        valid_insights = []
        for insight in data['insights'][:10]:
            if isinstance(insight, dict) and 'text' in insight and 'type' in insight:
                if insight['type'] not in ('learned', 'weakness'):
                    insight['type'] = 'learned'
                insight.setdefault('citations', [])
                valid_insights.append(insight)

        data['insights'] = valid_insights
        return data

    except (json.JSONDecodeError, KeyError, IndexError) as e:
        logger.error("[InsightExtractor] Failed to parse response: %s", e)
        return None

File: functions/src/handlers/insights.ts
import { Request, Response } from 'express';
import { createErrorResponse, ErrorCode } from '../utils/errors';
import { createLogger } from '../utils/logging';
import { debitTokens, getCurrentDateString, getCurrentWeekString } from '../utils/firestore';
import { calculateNormalizedTokens, calculateCostMicrodollars } from '../utils/tokenPricing';
import { chatCompletionWithRetry, resolveModel, OpenRouterRequest } from '../utils/openrouter';

// Fast, cheap model for structured extraction
const EXTRACTION_MODEL = 'gemini-2.5-flash';

const EXTRACTION_PROMPT = `Du extrahierst Lernerkenntnisse aus einem Chat über eine Anki-Lernkarte.

KARTE: {question}

CHAT:
{chat_messages}

FOKUS:
Erfasse was der NUTZER in diesem Chat gelernt, gefragt oder falsch verstanden hat.
Fasse NICHT den Tutor-Text zusammen — erfasse das Lernverhalten des Nutzers:

1. FRAGEN des Nutzers → zeigen Wissenslücken oder aktives Interesse (type: "learned")
2. FEHLER/VERWECHSLUNGEN → was der Nutzer falsch verstanden hat (type: "weakness")
3. UNSICHERHEITEN → wo der Nutzer nachfragen musste (type: "weakness")
4. SCHLÜSSELMOMENTE → was der Nutzer erst durch die Erklärung verstanden hat (type: "learned")

REGELN:
- Formuliere aus Nutzerperspektive: "Gefragt nach X", "Verwechslung: X vs Y", "Verstanden: X"
- Jede Frage des Nutzers ist eine potenzielle Erkenntnis (zeigt Interesse oder Wissenslücke)
- NUR kartenrelevante Erkenntnisse, kein Off-Topic/Smalltalk
- Max 10 Einträge
- NUR das JSON-Array ausgeben, KEIN anderer Text

BEISPIEL-OUTPUT:
[{"text":"Gefragt nach weiteren Ansätzen am Tuber ischiadicum","type":"learned"},{"text":"Verwechslung: M. pterygoideus lateralis vs. medialis","type":"weakness"},{"text":"Verstanden: Km ändert sich nur bei kompetitiver Hemmung","type":"learned"}]`;

interface InsightItem {
  type: 'learned' | 'weakness';
  text: string;
}

interface InsightsRequestBody {
  messages: Array<{ role: string; content: string }>;
  cardContext?: { question?: string; answer?: string };
}

/**
 * POST /insights/extract
 * Extracts learning insights from chat history using a fast model via OpenRouter.
 * Requires authentication.
 */
export async function insightsExtractHandler(
  req: Request,
  res: Response
): Promise<void> {
  const requestId = `req-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  const logger = createLogger(requestId);

  try {
    const userId = (req as any).userId;
    if (!userId) {
      res.status(401).json(
        createErrorResponse(ErrorCode.TOKEN_INVALID, 'Authentication required', undefined, requestId)
      );
      return;
    }

    // ── Validate body ────────────────────────────────────────────────
    const body: InsightsRequestBody = req.body;

    if (!body.messages || !Array.isArray(body.messages) || body.messages.length === 0) {
      res.status(400).json(
        createErrorResponse(ErrorCode.VALIDATION_ERROR, 'messages array is required and must not be empty', undefined, requestId)
      );
      return;
    }

    logger.info('Insights extraction requested', { userId, messageCount: body.messages.length });

    // ── Build extraction prompt ──────────────────────────────────────
    const question = (body.cardContext?.question || '').substring(0, 300);

    // Format chat messages into compact text
    const chatLines = body.messages
      .filter((m) => m.role === 'user' || m.role === 'assistant')
      .slice(-20) // Keep last 20 messages
      .map((m) => {
        const sender = m.role === 'user' ? 'User' : 'Tutor';
        const text = (m.content || '').substring(0, 400);
        return `${sender}: ${text}`;
      });
    const chatStr = chatLines.join('\n').substring(0, 3000);

    const prompt = EXTRACTION_PROMPT
      .replace('{question}', question)
      .replace('{chat_messages}', chatStr);

    // ── Call OpenRouter (non-streaming) ──────────────────────────────
    const resolvedModel = resolveModel(EXTRACTION_MODEL);

    const openrouterReq: OpenRouterRequest = {
      model: EXTRACTION_MODEL,
      messages: [
        { role: 'user', content: prompt },
      ],
      temperature: 0.3,
      max_tokens: 1000,
      stream: false,
    };

    const data = await chatCompletionWithRetry(openrouterReq, 1);

    // ── Extract response text ────────────────────────────────────────
    const responseText: string = data.choices?.[0]?.message?.content || '';

    // ── Parse JSON insights ──────────────────────────────────────────
    let insights: InsightItem[] = [];
    try {
      insights = parseInsightsResponse(responseText);
    } catch (parseError) {
      logger.warn('Failed to parse insights response, returning empty array', { responseText: responseText.substring(0, 200) });
      // Return empty array gracefully
    }

    // ── Debit tokens ─────────────────────────────────────────────────
    let tokenInfo: { used: number } | undefined;
    if (data.usage) {
      const inputTokens = data.usage.prompt_tokens || 0;
      const outputTokens = data.usage.completion_tokens || 0;
      const normalized = calculateNormalizedTokens(resolvedModel, inputTokens, outputTokens);
      const cost = calculateCostMicrodollars(resolvedModel, inputTokens, outputTokens);
      const date = getCurrentDateString();
      const week = getCurrentWeekString();

      debitTokens(userId, date, week, normalized, inputTokens, outputTokens, cost).catch((error) => {
        logger.error('Failed to debit tokens', error, { userId, normalized });
      });

      tokenInfo = { used: normalized };
    }

    logger.info('Insights extraction completed', { userId, insightCount: insights.length, tokenInfo });

    res.json({ insights, tokens: tokenInfo });
  } catch (error: any) {
    logger.error('Error in insights extract handler', error, { userId: (req as any).userId });

    if (res.headersSent) return;

    res.status(500).json(
      createErrorResponse(ErrorCode.BACKEND_ERROR, 'Failed to extract insights', error.message, requestId)
    );
  }
}

/**
 * Parse the AI response text into a validated insights array.
 * Handles markdown code fences and partial JSON gracefully.
 */
function parseInsightsResponse(responseText: string): InsightItem[] {
  let text = responseText.trim();

  // Strip markdown code fences
  if (text.includes('```json')) {
    text = text.split('```json')[1].split('```')[0].trim();
  } else if (text.includes('```')) {
    text = text.split('```')[1].split('```')[0].trim();
  }

  // Try to find a JSON array in the text
  const arrayMatch = text.match(/\[[\s\S]*\]/);
  if (arrayMatch) {
    text = arrayMatch[0];
  }

  // Also handle the wrapped format { "insights": [...] }
  const objectMatch = text.match(/\{[\s\S]*"insights"\s*:\s*(\[[\s\S]*\])[\s\S]*\}/);
  if (objectMatch) {
    text = objectMatch[1];
  }

  const parsed = JSON.parse(text);

  if (!Array.isArray(parsed)) {
    return [];
  }

  // Validate and normalize each insight
  const valid: InsightItem[] = [];
  for (const item of parsed.slice(0, 10)) {
    if (typeof item === 'object' && item !== null && typeof item.text === 'string' && item.text.length > 0) {
      const type = item.type === 'weakness' ? 'weakness' : 'learned';
      valid.push({ type, text: item.text });
    }
  }

  return valid;
}
